Export multi-tabbed grids to excel

Sigmonia opened the pull request:
#### Rationale
Item #10158: Multi-sheet export to Excel workbooks of multi-tab grids

#### Related Pull Requests
* [Components](https://github.com/LabKey/labkey-ui-components/pull/814)

#### Changes
* Added ExportTabsXlsxAction
* refactor ExcelWriter to simplify configuration

Sigmonia committed:
add new action to support export of multi-tab grids

labkey-matthewb committed:
encode backslash too.  Not 'needed', but seems like a good defensive programming strategy to protect against html encoded text being incorrectly embedded in a javascript.

labkey-matthewb commented on query/src/org/labkey/query/controllers/QueryController.java:
```diff
@@ -1672,6 +1672,118 @@ void _export(ExportQueryForm form, QueryView view) throws Exception
+    @Marshal(Marshaller.Jackson)
+    @RequiresPermission(ReadPermission.class)
+    @Action(ActionType.Export.class)
+    public static class ExportTabsXLSXAction extends ReadOnlyApiAction<ExportTabsForm>
```
I might nitpick about the name.  The "Tab"s bit is SM UI thing.  There are other uses that certainly have nothing to do with tabs.

labkey-matthewb reviewed (changes requested):
This seems like a pretty practical solution (e.g. more limited view specification than the full QueryForm).  I don't love the api Name.  Would love to see even a simple Junit smoke test.

labkey-matthewb commented on api/src/org/labkey/api/data/ExcelWriter.java:
```diff
@@ -508,13 +509,37 @@ public void write(OutputStream stream)
+     * Renders the sheet then writes out the workbook to supplied stream
+     * @param response to write out the file
+     */
     public void write(HttpServletResponse response)
```
With the refactor these write() methods are getting pretty overloaded/ambiguous.  Maybe rename with long descriptive names?  Like renderSheetAndWrite() etc?

Sigmonia replied:
done, kept deprecated method as it is used in several other repos

Sigmonia committed:
Bind query parameters versus using custom query form
Merge QueryView.AliasColumnExcelWriter --> ExcelWriter
Fix renamed columns not being used
Fix query filters not being applied

Sigmonia committed:
Merge remote-tracking branch 'origin/develop' into fb_excelTabs

Sigmonia committed:
Merge remote-tracking branch 'origin/develop' into fb_excelTabs

Sigmonia committed:
Rename/deprecate ExcelWriter.write method as renderSheetAndWrite to better self document

Sigmonia committed:
Update export action name and properties

labkey-matthewb commented on query/src/org/labkey/query/controllers/QueryController.java:
```diff
@@ -1672,6 +1673,113 @@ void _export(ExportQueryForm form, QueryView view) throws Exception
+                if (entry.getValue() instanceof JSONArray)
+                {
+                    String key = entry.getKey();
+                    props.put(key, ((JSONArray) entry.getValue()).getString(0));
```
check length()==1?  (or at least length() > 0)

labkey-matthewb reviewed (changes requested):
Looks good.  I would still like to see one simple junit test with MockRequest and maybe two sheets (maybe just query core schema tables)

Sigmonia committed:
Merge remote-tracking branch 'origin/develop' into fb_excelTabs

Sigmonia committed:
merge from develop

Sigmonia committed:
address:
Issue #45388: Workflows excel export
Issue #45391: Excel export filename does not match table name as expected

Sigmonia committed:
add jsonarray check to form binding

Sigmonia committed:
bump api version

Sigmonia committed:
Merge remote-tracking branch 'origin/develop' into fb_excelTabs
